packages/common/src/failure.ts

```diff
@@ -0,0 +1,424 @@
+import { coresdk } from '@temporalio/proto';
+import { DataConverter, arrayFromPayloads } from './converter/data-converter';
+
+export const FAILURE_SOURCE = 'NodeSDK';
+export type TimeoutType = coresdk.common.TimeoutType;
+export const TimeoutType = coresdk.common.TimeoutType;
+export type RetryState = coresdk.common.RetryState;
+export const RetryState = coresdk.common.RetryState;
+
+/**
+ * Represents failures that can cross Workflow and Activity boundaries.
+ *
+ * Only exceptions that extend this class will be propagated to the caller.
+ *
+ * **Never extend this class or any of its derivatives.** They are to be used by the SDK code
+ * only. Throw an instance {@link ApplicationFailure} to pass application specific errors between
+ * Workflows and Activities.
```

Comment:
Could just make this constructor private and then expose a `create` function to allow constructing them for testing purposes. That way the class can't be extended.